# JSON Mini-Project

This Jupyter Notebook contains the solutions for the JSON Mini-Project
which is part of the Springboard curriculum. 

### 1) Find the 10 countries with most projects

We begin by loading all the libraries we will be using, 
and reading the data as a pandas dataframe.

In [30]:
# import necessary libraries
import json
import pandas as pd
from pandas.io.json import json_normalize
import numpy as np

# read in json data
with open("data/world_bank_projects.json") as json_file:
    json_data = json.load(json_file)

# read data as dataframe, look at first 3 lines
world_bank = json_normalize(json_data)
world_bank.head(3)

_id.$oid                        ...                                                                        url
0  52b213b38594d8a2be17c780                        ...                          http://www.worldbank.org/projects/P129828/ethi...
1  52b213b38594d8a2be17c781                        ...                          http://www.worldbank.org/projects/P144674?lang=en
2  52b213b38594d8a2be17c782                        ...                          http://www.worldbank.org/projects/P145310?lang=en

[3 rows x 55 columns]

Next we'll count how many projects each country had, list the countries in descending order, and find the top ten.

In [31]:
# group data and count number of projects by country, display first 10
world_bank.groupby("countryshortname").size().sort_values(ascending = False).head(10)

countryshortname
Indonesia             19
China                 19
Vietnam               17
India                 16
Yemen, Republic of    13
Nepal                 12
Bangladesh            12
Morocco               12
Mozambique            11
Africa                11
dtype: int64

### 2) Find the top 10 major project themes (using column 'mjtheme_namecode')

Now we wish to find the top ten most common themes. This information is nested within the 'mjtheme_namecode' column. So we will need to extract the data into a new dataframe, and find the top ten themes.

In [32]:
# use json_normalize to access nested data
theme = json_normalize(data = json_data, record_path = "mjtheme_namecode")
theme.head(5)

code                                   name
0    8                      Human development
1   11                                       
2    1                    Economic management
3    6  Social protection and risk management
4    5                  Trade and integration

We can see we may have some missing data in the 'name' column, but we'll worry about this in the next problem. For now we'll go ahead and group the data by theme name, count how many there are, and display the top ten.

In [33]:
# group by theme name, display top ten most common themes
theme.groupby("name").size().sort_values(ascending = False).head(10)

name
Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
dtype: int64

### 3) Create a dataframe with the missing names filled in.

The above result doesn't make much sense, the theme with no name has 122 entries. So now we'll deal with those missing names, and then go back and update our previous result.

In [34]:
# create new dataframe 
df = theme.sort_values(by = "code", ascending = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1499 entries, 458 to 1102
Data columns (total 2 columns):
code    1499 non-null object
name    1499 non-null object
dtypes: object(2)
memory usage: 35.1+ KB


No missing values indicates we have empty strings, which we'll replace with NaN values. We can then replace those with the correct strings.

In [37]:
# replace empty string with NaN
df.replace(to_replace = "", value = np.nan, inplace = True)

# confirm we replaced empty strings with NaN
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1499 entries, 458 to 1102
Data columns (total 2 columns):
code    1499 non-null object
name    1499 non-null object
dtypes: object(2)
memory usage: 35.1+ KB


In [39]:
# replace NaN with correct string
df["name"].fillna(inplace = True, method = "bfill")

# check that NaN were replaced
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1499 entries, 458 to 1102
Data columns (total 2 columns):
code    1499 non-null object
name    1499 non-null object
dtypes: object(2)
memory usage: 35.1+ KB


Now that we've replaced the empty names with their correct names, we'll once again find the top ten themes.

In [42]:
# with new df, find top ten themes
df.groupby(["code", "name"]).size().sort_values(ascending = False).head(10)

code  name                                        
11    Environment and natural resources management    250
10    Rural development                               216
8     Human development                               210
2     Public sector governance                        199
6     Social protection and risk management           168
4     Financial and private sector development        146
7     Social dev/gender/inclusion                     130
5     Trade and integration                            77
9     Urban development                                50
1     Economic management                              38
dtype: int64

In [43]:
# double check this is correct by counting the top ten "code" variable
df.groupby("code").size().sort_values(ascending = False).head(10)

code
11    250
10    216
8     210
2     199
6     168
4     146
7     130
5      77
9      50
1      38
dtype: int64